vmaas_sync/metrics_cyndi.go

```diff
@@ -0,0 +1,163 @@
+		Namespace: "patchman_engine",
+		Subsystem: "vmaas_sync",
+		Name:      "cyndi_systems",
+	}, []string{"type"})
+
+	cyndiTagsCnt = prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Namespace: "patchman_engine",
+		Subsystem: "vmaas_sync",
+		Name:      "cyndi_tags_count",
+	}, []string{"type"})
+)
+
+type InventoryHosts []struct {
+	DisplayName   string
+	Tags          string
+	SystemProfile string
+}
+
+//Firstly, this method unmarshal jsonb fields from DB, and loads them into relevant struct
+//This method also exposes metrics regarding max/min tags per system and number of systestems
+//with/without tags and counts systems with sap_system=true
+func updateCyndiData() {
+	count, err := getCyndiData()
+	if err != nil {
+		utils.Log("err", err.Error()).Error("unable to update cyndi metrics")
+		return
+	}
+
+	for key, value := range count {
+		cyndiTagsCnt.WithLabelValues(key).Set(float64(value))
+	}
+}
+
+//Loads display_name, tags and system_profile from inventory.hosts
+//and loads them in InventoryHosts struct.
+//Tags are parsed into from jsonb string to TagSlice struct, which contains struct Tag with individual tags.
+//All tags are then stored in uniqueTagsSlice variable to compute, which tags are unique
+func getCyndiData() (map[string]int, error) {
+	count := map[string]int{}
+	uniqueTagsSlice := []models.TagSlice{}
+	var stats InventoryHosts
+	err := database.Db.Table("inventory.hosts").Scan(&stats).Error
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to get inventory hosts from db")
+	}
+
+	var systemsWithTag int = 0
+	var systemsSapSystem int = 0
+	count["allSystemCount"] = len(stats)
+	for _, item := range stats {
+		var tags models.TagSlice
+		err := json.Unmarshal([]byte(item.Tags), &tags)
+		if err != nil {
+			utils.Log("err", err.Error()).Error("Not able to unmarshall")
+		}
+		uniqueTagsSlice = append(uniqueTagsSlice, tags)
+		var systemProfile models.SystemProfile
+		err = json.Unmarshal([]byte(item.SystemProfile), &systemProfile)
+		if err != nil {
+			utils.Log("err", err.Error()).Error("Not able to unmarshall")
+		}
+		if tags != nil {
+			systemsWithTag++
+			maxTag := len(tags)
+			count[item.DisplayName] = maxTag
+		}
+		if systemProfile.SapSystem {
+			systemsSapSystem++
+		}
+	}
+	uniqueTags := 0
+	for _, value := range unique(uniqueTagsSlice) {
+		if value {
+			uniqueTags++
+		}
+	}
+	count["uniqueTagsCount"] = uniqueTags
+	count["systemsSapSystem"] = systemsSapSystem
+	count["systemsWithTag"] = systemsWithTag
+
+	return count, nil
+}
+
+//Adds all tags to map, and checks, if any of these tags is unique
+//Uses method contains, which returns true, if tag is already in map's keys
+//Therefore value returned from method must be negated
+func unique(inputTags []models.TagSlice) map[models.Tag]bool {
+	mapOfUniqueTags := make(map[models.Tag]bool)
+
+	for _, singleTagSlice := range inputTags {
+		for _, singleTag := range singleTagSlice {
+			mapOfUniqueTags[singleTag] = !contains(mapOfUniqueTags, singleTag)
```

Comment:
You don't have to use reflection here. The expression `mapOfUniqueTags[tag]` will return false if `tag` is not contained in the map